asf_levies_model/consumers.py

```diff
@@ -0,0 +1,460 @@
+        consumer_from_dataframe():
+            Creates a Consumer instance from a dataframe input.
+        boiler_to_hp_savings():
+            Returns a cost savings estimate if consumer is assumed to switch from a gas boiler for heat to an electric heat pump.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        archetype: str,
+        net_annual_income: float,
+        net_income_decile: int,
+        main_heating_fuel: str,
+        gas_consumption: float,
+        electricity_consumption: float,
+        gas_tariff: object,
+        electricity_tariff: object,
+        unmetered_fuel_spend: float = 0,
+        scheme_eligible: bool = False,
+        size: int = 0,
+    ) -> None:
+        """Initalizes a Consumer instance based on provided parameters.
+
+        Parameters
+        ----------
+        name : str
+            Unique name of the energy consumer.
+        archetype : str
+            Name of broader archetype the consumer belongs to.
+        net_annual_income : float
+            The average net annual income of represented consumer, in £.
+        net_income_decile : int
+            The income decile that the consumer's net annual income falls in.
+        main_heating_fuel : str
+            The name of the consumer's main heating fuel.
+        gas_consumption : float
+            Annual gas consumption, in MWh.
+        electricity_consumption : float
+            Annual electricity consumption, in MWh.
+        gas_tariff : object
+            Tariff object to be used to calculate the consumer's gas bill.
+        electricity_tariff : object
+            Tariff object to be used to calculate the consumer's electricity bill.
+        unmetered_fuel_spend : float, optional
+            Annual amount spent on unmetered fuel (i.e., fuel that is not gas or electricity), in £, by default 0.
+        scheme_eligible : bool, optional
+            Eligibilty for social support scheme, by default False.
+        size : int, optional
+            Number of households in consumer profile, by default 0 (i.e. size is not considered).
+        """
+
+        self.name = name
+        self.archetype = archetype
+        self.net_annual_income = net_annual_income
+        self.net_income_decile = net_income_decile
+        self.main_heating_fuel = main_heating_fuel
+        self.gas_consumption = gas_consumption
+        self.electricity_consumption = electricity_consumption
+        self.scheme_eligible = scheme_eligible
+        self.unmetered_fuel_spend = unmetered_fuel_spend
+        self.size = size
+
+        self.electricity_tariff = electricity_tariff
+        self.gas_tariff = gas_tariff
+
+        self._electricity_bill = None
+        self._gas_bill = None
+
+        self._adjustment_mode = None
+
+    @property
+    def unit_cost_ratio(self) -> float:
+        if self._adjustment_mode == "unit discount":
```

Comment:
I think this is a good start, however I think we need to think a little more about the adjustment.

We define the unit cost ratio as the (pre-VAT) cost of 1 unit of electricity / 1 unit of gas.

However, the adjustment at the moment (I think) is the post-VAT difference in total (variable + fixed) cost of energy normalised by consumption.

In a way, this muddies the waters between having a tariff-based spark gap and a modified tariff+support-based spark gap. However this approach means that the adjusted spark gap is consumer specific by the ratio of standing charge to consumption. I guess in some circumstances, where rebates/support specifically target the unit cost of consumption this kind of approach makes sense, however others such as in the case of a fixed rebate it's less clear.